hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionManager.java

```diff
@@ -139,26 +142,77 @@ private List<DatanodeDetails> mapHostnamesToDatanodes(List<String> hosts)
               + " Please check the port number.");
         }
         results.add(found.get(0));
-      } else if (found.size() > 1) {
-        DatanodeDetails match = null;
-        for (DatanodeDetails dn : found) {
-          if (validateDNPortMatch(host.getPort(), dn)) {
-            match = dn;
-            break;
+      } else {
+        // Here we either have multiple DNs on the same host / IP, and they
+        // should have different ports. Or, we have a case where a DN was
+        // registered from a host, then stopped and formatted, changing its
+        // UUID and registered again. In that case, the ports of all hosts
+        // should be the same, and we should just use the one with the most
+        // recent heartbeat.
+        if (allPortsMatch(found)) {
+          // All ports match, so just use the most recent heartbeat as it is
+          // not possible for a host to have 2 DNs coming from the same port.
+          DatanodeDetails mostRecent = findDnWithMostRecentHeartbeat(found);
+          if (mostRecent == null) {
+            throw new InvalidHostStringException("Host " + host.getRawHostname()
+                + " has multiple datanodes registered with SCM."
+                + " All have identical ports, but none have a newest"
+                + " heartbeat.");
           }
+          results.add(mostRecent);
+        } else {
+          DatanodeDetails match = null;
+          for (DatanodeDetails dn : found) {
```

Comment:
Here can we have the below possibility? 
Let's say it matches 3 DNs in which 2 DN are having same port but not the third one.
And assume` host port `matches the 2 DN.
In this case it will not able to choose which one to decommission and here also we need `findDnWithMostRecentHeartbeat` among the 2 DN
Can you consider `host port` also in `allPortsMatch` so that this condition will not arise?

Reply:
I have added some logic to handle these sort of cases. I think have multiple registrations from the same host and different ports on a "real" cluster would be very unlikely, but we may as well fix it while we are here.